mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp

```diff
@@ -243,7 +187,168 @@ FlatSymbolRefAttr ZneLowering::getOrInsertFoldedCircuit(Location loc, PatternRew
+
+    // Function without measurements: Create function without measurements and with qreg as last
+    // argument
+    func::FuncOp fnWithoutMeasurementsOp;
+    func::FuncOp fnWithMeasurementsOp;
+    if (foldingAlgorithm == Folding(1)) {
+        FlatSymbolRefAttr fnWithoutMeasurementsRefAttr =
+            getOrInsertFnWithoutMeasurements(loc, rewriter, op);
+        fnWithoutMeasurementsOp =
+            SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, fnWithoutMeasurementsRefAttr);
+        // Function with measurements: Modify the original function to take a quantum register as
+        // last arg and keep measurements
+        FlatSymbolRefAttr fnWithMeasurementsRefAttr =
+            getOrInsertFnWithMeasurements(loc, rewriter, op);
+        fnWithMeasurementsOp =
+            SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, fnWithMeasurementsRefAttr);
+    }
+    else {
+    }
+    rewriter.setInsertionPointToStart(moduleOp.getBody());
+
+    FunctionType fnFoldedType = FunctionType::get(ctx, /*inputs=*/
+                                                  typesFolded,
+                                                  /*outputs=*/fnOp.getResultTypes());
+
+    func::FuncOp fnFoldedOp = rewriter.create<func::FuncOp>(loc, fnFoldedName, fnFoldedType);
+    Value c0, c1, numberQubitsValue;
+    fnFoldedOp.setPrivate();
+    if (foldingAlgorithm == Folding(1)) {
+        Block *fnFoldedOpBlock = fnFoldedOp.addEntryBlock();
+        rewriter.setInsertionPointToStart(fnFoldedOpBlock);
+        // Loop control variables
+        c0 = rewriter.create<index::ConstantOp>(loc, 0);
+        c1 = rewriter.create<index::ConstantOp>(loc, 1);
+        TypedAttr numberQubitsAttr = rewriter.getI64IntegerAttr(numberQubits);
+        numberQubitsValue = rewriter.create<arith::ConstantOp>(loc, numberQubitsAttr);
+        rewriter.create<quantum::DeviceInitOp>(loc, lib, name, kwargs);
+    }
+    else {
+        rewriter.cloneRegionBefore(fnOp.getBody(), fnFoldedOp.getBody(), fnFoldedOp.end());
+
+        Block *fnFoldedOpBlock = &fnFoldedOp.getBody().front();
+        rewriter.setInsertionPointToStart(fnFoldedOpBlock);
+        // Loop control variables
+        c0 = rewriter.create<index::ConstantOp>(loc, 0);
+        c1 = rewriter.create<index::ConstantOp>(loc, 1);
+
+        fnFoldedOpBlock->addArgument(fnFoldedOp.getArgumentTypes().front(), loc);
+    }
+
+    if (foldingAlgorithm == Folding(1)) {
+        return globalFolding(loc, rewriter, fnFoldedName, lib, name, kwargs, fnFoldedType,
+                             typesFolded, fnFoldedOp, fnAllocOp, numberQubitsValue,
+                             fnWithoutMeasurementsOp, fnWithMeasurementsOp, c0, c1);
+    }
+    if (foldingAlgorithm == Folding(2)) {
+        return randomLocalFolding(rewriter, fnFoldedName, fnFoldedOp, c0, c1);
+    }
+    // Else, if (foldingAlgorithm == Folding(3)):
+    return allLocalFolding(rewriter, fnFoldedName, fnFoldedOp, c0, c1);
```

Comment:
Switch the folding if random is not implemented in this PR.

Reply:
Done.